### PR TITLE
Add MP2 Gradients

### PR DESCRIPTION
**Pull request by kcpearce**
## Description
This PR adds functionality for nuclear gradients at the MP2 level.

## What are your new additions? Please provide a brief list.
* **New Features**
  - [x] MP2 Gradients

* **Changes**
  - [x] Updates README to include MP2_Gradient.py short description 

## Status
- [ ] Click when ready for review-and-merge

### CODE REVIEW

**Comment by kcpearce** [comment]
@dsirianni I see that the Travis CI build is failing and it appears to have to do with the RHF_Gradient and RHF_Hessian tests. It seems as though it has to do with some convergence criteria since the results match to a certain number of decimal places. However, I don't think I made any changes to these files, so I'm not sure why they wouldn't pass. Any idea?

**Comment by kcpearce** [comment]
Perhaps @ashutoshvt has some insight?

**Comment by loriab** [comment]
I suspect it's another casualty of the physical constants update to psi4. bohr2ang changed slightly, so nuclear repulsion energies shift accordingly when test mols are in Angstrom. It shows up now b/c travis hasn't run in a while b/c no new commits.

File: Moller-Plesset/test.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import time
```
Comment (amjames) [inline Moller-Plesset/test.py]: You will want to `git rm ` this file

File: Self-Consistent-Field/Untitled.ipynb [hunk 2]
```diff
@@ -0,0 +1,296 @@
+{
```
Comment (amjames) [inline Self-Consistent-Field/Untitled.ipynb]: You will want to `git rm ` this file

File: Self-Consistent-Field/Untitled1.ipynb [hunk 3]
```diff
@@ -0,0 +1,76 @@
+{
```
Comment (amjames) [inline Self-Consistent-Field/Untitled1.ipynb]: You will want to `git rm` this file

**Comment by ashutoshvt** [comment]
I agree with @loriab:

**Comment by ashutoshvt** [comment]
So, the reference values need to be updated accordingly.

File: Moller-Plesset/MP2_Gradient.py [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+        Gradient["J"][atom, p] += 2.0 * np.einsum('pq,pqmm->', ref_opdm, deriv1_np[map_key][:,:,:nocc,:nocc])
+        Gradient["K"][atom, p] -= 1.0 * np.einsum('pq,pmmq->', ref_opdm, deriv1_np[map_key][:,:nocc,:nocc,:])
+
+        # Reference TPDM component of TEI gradients
+        Gradient["J"][atom, p] += 4.0 * np.einsum('pqrs,prqs->', ref_tpdm, deriv1_np[map_key])
+
+        # Contract MP2 TPDM altogether
+        Gradient["J"][atom, p] += 4.0 * np.einsum('pqrs,prqs->', Ppqrs, deriv1_np[map_key])
+
+        # Contract MP2 TPDM by individual components
```
Comment (amjames) [inline Moller-Plesset/MP2_Gradient.py]: If these commented blocks are potentially useful then leave them, but if not you can remove them. May want to see what @dgasmith, @dsirianni have to say

File: Self-Consistent-Field/Untitled.ipynb (hunk 2 above)
Comment (kcpearce) [reply Self-Consistent-Field/Untitled.ipynb]: Done.

File: Self-Consistent-Field/Untitled1.ipynb (hunk 3 above)
Comment (kcpearce) [reply Self-Consistent-Field/Untitled1.ipynb]: Done.

File: Moller-Plesset/MP2_Gradient.py (hunk 4 above)
Comment (kcpearce) [reply Moller-Plesset/MP2_Gradient.py]: Yeah, these might be helpful to me in the future when I'm working on things on Psi4 side of things, where I'll have to implement each individual block of the TPDM separately. But I imagine I could probably get rid of them here.

File: Moller-Plesset/MP2_Gradient.py [hunk 5]
```diff
@@ -0,0 +1,506 @@
...
+# where t2_tilde = [2<ij|ab> - <ij|ba>] / (e_i + e_j - e_a - e_b)
+t2_tilde = 2 * t2 - t2.swapaxes(2,3)
+
+
+# Build Reference OPDM
+D_ao = wfn.Da()
+D_ao.add(wfn.Db())
+npD_ao = D_ao.to_array()
+# Transform to MO Basis
+ref_opdm = np.einsum('iu,uv,vw,wx,xj', npC.T, npS.T, npD_ao, npS, npC)
```
[the remark this answers is not in the document]
Comment (kcpearce) [reply Moller-Plesset/MP2_Gradient.py]: Should I use it on einsum calls throughout the entire script? Or is it just especially necessary for this call?

**Comment by dgasmith** [comment]
For Python2 it looks like you have a non-ASCII char. You can add the following to the top line to correct this: `# -*- coding: utf-8 -*-`

**Comment by kcpearce** [comment]
@dgasmith @loriab It looks like Travis CI is still failing on tests `1g_basis-sets`, `4a_DFT_Grid`, `4b_LDA_kernel`, `4d_VV10`, and `4e_GRAC` with very similar errors. Not exactly sure what's causing it.

**Comment by kcpearce** [comment]
@loriab I saw that [psi4/psi4#1339](https://github.com/psi4/psi4/pull/1339) was supposed to fix 4b, 4d, and 4e py36 failing tests, but it seems they're still failing. Thoughts?

**Comment by loriab** [comment]
That fix has to work it’s way into a conda build first, then tests will have to be retriggered here.

**Comment by alenaizan** [comment]
@dgasmith It appears one charge has an error of 1.315e-05, which is slightly larger than the tolerance but it is still a reasonable error. I think we should loose up the check tolerance. It seems the charge fitting might have some reproducibility issue beyond maybe 4-5 digits. I tried the test in a local machine using both py2.7 and py3.6 and they both ran successfully.
By the way, when do you expect to drop support for py2.7 in psi4numpy?

File: Self-Consistent-Field/RHF_Gradient.py [hunk 6]
```diff
@@ -161,9 +161,9 @@
...
 Total_G_psi4 = psi4.core.Matrix.from_list([                                     
-             [ 0.000000000000, -0.000000000000, -0.097441440379], 
-             [-0.000000000000, -0.086300100260,  0.048720720189],
-             [-0.000000000000,  0.086300100260,  0.048720720189]
-       ])
+        [ 0.000000000000, -0.000000000000, -0.097441440379],
+        [-0.000000000000, -0.086300100260,  0.048720720189],
+        [-0.000000000000,  0.086300100260,  0.048720720189]
+    ])
 G_python_Total_mat = psi4.core.Matrix.from_array(Gradient["Total"])
```
[the remark this answers is not in the document]
Comment (kcpearce) [reply Self-Consistent-Field/RHF_Gradient.py]: To clarify, is "G_python_total_mat" what you're looking for here?

File: Moller-Plesset/MP2_Gradient.py [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+Gradient["TEI"] = 0.25 * (Gradient["J"] + Gradient["K"])
+Gradient["Total"] = Gradient["OEI"] + Gradient["TEI"] + Gradient["N"]
+
+psi4.core.print_out("\n\n Total Gradient:\n\n")
+Tot_grad = psi4.core.Matrix.from_array(Gradient["Total"])
+Tot_grad.name = " TOTAL GRADIENT"
+Tot_grad.print_out()
+
+# Psi4's Total Gradient
+Psi4_total_grad = psi4.core.Matrix.from_list([
```
[the remark this answers is not in the document]
Comment (kcpearce) [reply Moller-Plesset/MP2_Gradient.py]: And is "psi4_total_grad" what you're looking for here?

**Comment by loriab** [comment]
for testing we try to make sure p4n works against the last release, so after 1.3, p4n can be py36+

**Comment by dgasmith** [comment]
@kcpearce Can you lower the check threshold slightly on the RESP test case? I do not think it is a bug, more of an overzealous threshold.

**Comment by kcpearce** [comment]
@dgasmith It appears like the RESP case was passing Travis tests, but I tried it locally and it failed. I went ahead and loosened the threshold slightly to get it to pass locally. It is still passing Travis, but there are still a few tests failing py36. It appears that [psi4/psi4#1339](https://github.com/psi4/psi4/pull/1339) was supposed to fix these test, but perhaps it hasn't made it into a conda build yet. Any idea how long that process should take?

**Comment by loriab** [comment]
It helps that I just turned the nightly build script back to psi4 from helPME. Running psi4 now. I'd rather keep p4n working with latest stable release (py27) _and_ dev head (py36).

**Comment by loriab** [comment]
you're green!

**Pull request merged**